File: 백준/Gold/4179. 불！/불！.py
from collections import deque

R, C = map(int, input().split())

miro = [list(input()) for _ in range(R)]
moveX = [-1, 1, 0, 0]
moveY = [0, 0, -1, 1]

jQueue = deque()
fQueue = deque()

jVisited = [[0] * C for _ in range(R)]
fVisited = [[0] * C for _ in range(R)]

answer = 0

def bfs():
    while fQueue:
        x, y = fQueue.popleft()

        for i in range(4):
            newX = x + moveX[i]
            newY = y + moveY[i]

            if 0 <= newX < R and 0 <= newY < C:
                if fVisited[newX][newY] != 0 or miro[newX][newY] == '#': continue

                fVisited[newX][newY] = fVisited[x][y] + 1
                fQueue.append((newX, newY))
    
    while jQueue:
        x, y = jQueue.popleft()

        for i in range(4):
            newX = x + moveX[i]
            newY = y + moveY[i]

            if 0 <= newX < R and 0 <= newY < C:
                if jVisited[newX][newY] != 0 or miro[newX][newY] == '#': continue
                if jVisited[x][y] + 1 < fVisited[newX][newY] or fVisited[newX][newY] == 0:
                    jVisited[newX][newY] = jVisited[x][y] + 1
                    jQueue.append((newX, newY))
            else: return jVisited[x][y]

    return "IMPOSSIBLE"

for i in range(R):
    for j in range(C):
        if miro[i][j] == 'J':
            jQueue.append((i, j))
            jVisited[i][j] = 1
        elif miro[i][j] == 'F':
            fQueue.append((i, j))
            fVisited[i][j] = 1

print(bfs())

File: 백준/Gold/4179. 불！/불！.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static char[][] board;
    static int[][] fireTime;
    static int[][] jihoonTime;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int r;
    static int c;
    static Queue<int[]> JQueue = new LinkedList<>();
    static Queue<int[]> FQueue = new LinkedList<>();

    private static void bfs() {
        while (!FQueue.isEmpty()) {
            int[] temp = FQueue.poll();
            int x = temp[0];
            int y = temp[1];

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (0 <= newX && newX < r && 0 <= newY && newY < c) {
                    if (board[newX][newY] == '.' && fireTime[newX][newY] == 0) {
                        FQueue.offer(new int[]{newX, newY});
                        fireTime[newX][newY] = fireTime[x][y] + 1;
                    }
                }
            }
        }

        while (!JQueue.isEmpty()) {
            int[] temp = JQueue.poll();
            int x = temp[0];
            int y = temp[1];

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (newX < 0 || newX >= r || newY < 0 || newY >= c) {
                    System.out.println(jihoonTime[x][y]);
                    return;
                }

                if (board[newX][newY] == '.' && jihoonTime[newX][newY] == 0 && (fireTime[newX][newY] == 0 || jihoonTime[x][y] + 1 < fireTime[newX][newY])) {
                    JQueue.offer(new int[]{newX, newY});
                    jihoonTime[newX][newY] = jihoonTime[x][y] + 1;
                }
            }
        }

        System.out.println("IMPOSSIBLE");
    }

    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stringTokenizer = new StringTokenizer(bufferedReader.readLine());
        r = Integer.parseInt(stringTokenizer.nextToken());
        c = Integer.parseInt(stringTokenizer.nextToken());

        board = new char[r][c];
        fireTime = new int[r][c];
        jihoonTime = new int[r][c];

        for (int i = 0; i < r; i++) {
            String temp = bufferedReader.readLine();
            for (int j = 0; j < c; j++) {
                board[i][j] = temp.charAt(j);
            }
        }

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (board[i][j] == 'J') {
                    JQueue.offer(new int[]{i, j});
                    jihoonTime[i][j] = 1;
                }
                if (board[i][j] == 'F') {
                    FQueue.offer(new int[]{i, j});
                    fireTime[i][j] = 1;
                }
            }
        }

        bfs();
    }
}
